Implement linked-list `LocalChain` and add rpc-chain module/example

evanlinjin opened the pull request:
Fixes bitcoindevkit/bdk#997 

### Description

This PR changes the `LocalChain` implementation to have blocks stored as a linked-list. To verify the API of the linked-list `LocalChain`, the `bitcoind_rpc` crate and example is added (which is a block-by-block chain source). Additionally, the electrum/esplora crates and examples are also updated for the new API, but the logic is also updated to better-ensure consistency when obtaining a view of blocks.

A linked-list `LocalChain` allows the data-source thread to hold a shared reference to a single checkpoint and have access to the whole history of checkpoints without cloning or having a lock on `LocalChain`. This idea was originally proposed by @LLFourn's.

### Notes to the reviewers

Sorry this is a massive PR again. The changes can be "grouped" as follows:
* Change `LocalChain` API to use linked-list.
  * The `apply_update` method is renamed to `update` and the input is now the `CheckPoint` tip. An additional input is added (`introduce_older_blocks: bool`) which allows the caller to specify whether they wish to introduce new checkpoints below the highest point of agreement. This is required optimization as the update tip's chain and the original chain may share a large portion of history. 
  * A `_check_consistency` method is added, which `assert!`s whether the internal tip's history is consistent with the internal index.
* Introduce `prune_and_apply_update` methods for both `IndexedTxGraph` and `Wallet`. This prunes out irrelevant transactions from the `TxGraph` update before applying it. This will be used by block-by-block data sources as each block update is a `TxGraph`.
* Update `electrum` crate to use the new `LocalChain` API. We cannot 100% guarantee consistency with `electrum` updates (yet), but this is an improvement.
* Update `esplora` to ensure 100% consistency. `LocalChain` and `TxGraph` updates are now handled separately in `ExploraExt`. We make use of the `block_hash` field of tx statuses from the esplora API.
* Introduce `bitcoind_rpc` crate and example.

### Changelog notice

* Change `LocalChain` API to use a linked-list of block data.
* Add the `bitcoind_rpc` crate and example.
* Fix the `esplora` crate to guarantee a consistent view of blocks when formulating an update.
* Add `example_rpc` CLI example for `bitcoind_rpc` crate.
* Fix: Remove duplicate temp-plan crate.

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [x] I've added docs for the new feature

#### Bugfixes:

* [x] This pull request breaks the existing API
~* [ ] I've added tests to reproduce the issue which are now passing~
* [x] I'm linking the issue being fixed by this PR

evanlinjin committed:
[local_chain] Implement `LocalChain` with linked list

This allows the data-source thread to hold a reference to checkpoints
without a lock on `LocalChain` itself.

Introduce `LocalChain::update` that replaces `determine_changeset` and
`apply_update`. This returns a closure that actually updates `self` when
called. This method allows for efficient and elegant updating while
being able to "preview" the update before applying.

The `LocalChain` update/determine_changeset tests have been updated to
also check for the final state after applying the update (not just
looking at the changeset).

Update `keychain::LocalUpdate` struct to use `CheckPoint`

Instead of containing a complete `LocalChain`, the update uses
`CheckPoint`. This simplifies the API since updating a `LocalChain` only
requires a `CheckPoint` now.

The examples and chain source `..Ext` implementations have all been
updated to use the new API.

Additionally, `..Ext` implementations didn't 100% guarantee consistency
of the updates, the logic has been changed to enforce better guarantees.

Commit:
Add esplora CLI example

LLFourn commented on crates/chain/src/local_chain.rs:
```diff
@@ -1,14 +1,116 @@
+    /// Block height.
+    pub height: u32,
+    /// Block hash.
+    pub hash: BlockHash,
```
Why not just hacve `block_id` here instead of these two fields.

evanlinjin replied:
No specific reason, just that it seems more convenient this way if we are exposing `CheckPoint`. I.e. you can get blockhash by `cp.hash` instead of `cp.block.hash`.

evanlinjin replied:
Changed.

LLFourn commented on crates/chain/src/local_chain.rs:
```diff
@@ -1,14 +1,116 @@
+}
+
+impl Iterator for CheckPointIter {
+    type Item = Arc<CheckPoint>;
```
Why doesn't this just iterate `BlockId`.

evanlinjin replied:
My thinking is that `CheckPoint` just contains more data, so why not just expose that?

evanlinjin replied:
@LLFourn and myself discussed about this face-to-face and we agree that iterating over checkpoints is a useful feature to have since we can compare raw pointers, making updates more efficient. However, `Arc` should be an internal implementation detail.

These changes have all been implemented.

LLFourn commented on crates/chain/src/local_chain.rs:
```diff
@@ -17,224 +119,450 @@ impl ChainOracle for LocalChain {
+    pub fn determine_changeset(
+        &self,
+        new_tip: &Arc<CheckPoint>,
+        must_include_height: Option<u32>,
```
why isn't the update another `LocalChain`?

evanlinjin replied:
Formulating a `LocalChain` means we need to think about which checkpoints to include in the update. I think it's a much easier API if the chain-data-source-thread just spits out new checkpoints it is aware of, and the `determine_changeset`. Of course, you can continue to add data to an update `LocalChain` and throw that down a channel, but the internal `BTreeMap` will keep growing.

LLFourn replied:
Ok I think that makes sense. so what is the point of `must_include_height`? The doc comments don't mention it.

evanlinjin replied:
Removed `must_include_height`. We break early if the internal raw pointers match. Otherwise we iterate over the entire ancestry of `new_tip`.

LLFourn commented on crates/chain/src/local_chain.rs:
```diff
@@ -1,14 +1,116 @@
+
+/// Represents a block of [`LocalChain`].
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct CheckPoint {
```
This type should be private in my mind. It's an implementation detail for how you can produce an iterator for `BlockId` with a single `Arc::clone`.

evanlinjin replied:
What are the downsides of exposing `Arc`?

evanlinjin replied:
`Arc` is now hidden.

LLFourn commented on crates/chain/src/local_chain.rs:
```diff
@@ -1,14 +1,116 @@
+    checkpoints: BTreeMap<u32, Arc<CheckPoint>>,
+}
+
+impl FromIterator<Arc<CheckPoint>> for LocalChain {
```
Not sure what this is for. You can create a `LocalChain` from a single `&CheckPoint` by going backwards. Why do you want to create a localchain from an iterator of `CheckPoint`. Also since I think `CheckPoint` should be private I also think this probably shouldn't exist.

evanlinjin replied:
Removed.

LLFourn commented on crates/chain/src/local_chain.rs:
```diff
@@ -1,14 +1,116 @@
+    }
+}
+
+impl IntoIterator for LocalChain {
```
I guess we don't want this because it consumes `LocalChain` so is not very useful given how we expect to use it.

evanlinjin replied:
Removed.

LLFourn commented on crates/chain/src/local_chain.rs:
```diff
@@ -17,224 +119,450 @@ impl ChainOracle for LocalChain {
+    ///
+    /// The blocks do not need to be in order. If multiple blocks of the same height are introduced,
+    /// the latest block will be used.
+    pub fn from_blocks(blocks: impl IntoIterator<Item = BlockId>) -> Self {
```
Not sure about this API. Can't we just force them to pass in a `BTreeSet<BlockId>` so it's ordered properly and so we can set `prev` in a natural way.

The docs must say that it is up to the caller to make sure the blocks are in the same chain.

evanlinjin replied:
Changed as suggested.

LLFourn reviewed (commented):
Initial comments. The main difference I have is that the user should not see `Arc` anywhere. It's an internal implementation detail.

LLFourn commented on crates/chain/src/local_chain.rs:
```diff
@@ -1,14 +1,149 @@
+    }
+
+    /// Construct a [`CheckPoint`] of `block` with a previous checkpoint.
+    pub fn new_with_prev(
```
Do we need this as a public API?

What about if you can just create a `CheckPoint` from a `BTreeSet` of `BlockId` which are assumed to be in the same chain. This would remove the error case that comes with this method.

evanlinjin replied:
Ideally we want to create with a previous pointer. This way updates from the data source can connect in memory.

LLFourn replied:
What about a method on checkpoint like extend which creates a new one on top of an existing one?

evanlinjin replied:
Heights may not increment in sequence, so don't we still need an error? We can also have a Result, where the error variant is the old checkpoint (meaning failed to extend).

LLFourn replied:
Yes sorry this wouldn't remove the error case but just make it a more natural API (pushing onto the end of a list).

evanlinjin replied:
Changed the API to what you suggested @LLFourn

LLFourn commented on crates/chain/src/local_chain.rs:
```diff
@@ -17,224 +152,339 @@ impl ChainOracle for LocalChain {
+    /// # Errors
+    ///
+    /// Replacing the block hash of an existing checkpoint will result in an error.
+    pub fn get_or_insert(
```
What's wrong with `get_block_hash` and `insert_block`. Why does this API need to exist?

evanlinjin replied:
Reverted back to `insert_block`!

LLFourn commented on crates/chain/src/local_chain.rs:
```diff
@@ -17,224 +152,339 @@ impl ChainOracle for LocalChain {
 
-    /// This is like the sparsechain's logic, expect we must guarantee that all invalidated heights
-    /// are to be re-filled.
-    pub fn determine_changeset(&self, update: &Self) -> Result<ChangeSet, UpdateNotConnectedError> {
```
Don't you still need this? You want to `determine_changeset`, apply to persistent storage and then update in memory if it succeeds. In order to only have `update` we'd have to make `LocalChain` monotone. We could do this after this PR pretty easily I think and so leave it as `update` if we promise to fix it later.

evanlinjin replied:
`update` returns a closure first. The local chain only updates when the returned closure is called. It is done in this way so we can make use of the pointers during the update.

LLFourn replied:
I don't really like this. I can just call `apply_changeset` instead of calling this closure to do the same thing?. It'd be fine with just applying the changeset right away and trying to fix #1005 eventually.

evanlinjin replied:
Yes, but you won't get some of the benefits of the linked list. What do you not like about this API?

LLFourn replied:
It's redundant which makes it confusing. Is the point I might decide not to apply it? In which situation would I not want to apply it?

evanlinjin replied:
Ah I see what you mean, I'll remove the closure and have the method apply changes right away.

evanlinjin replied:
Done.

evanlinjin committed:
[bitcoind_rpc] Initial work on `BitcoindRpcIter`

LLFourn commented on crates/esplora/src/blocking_ext.rs:
```diff
@@ -66,60 +62,75 @@ pub trait EsploraExt {
 
+        let tip = last_cp.expect("must have atleast one checkpoint");
+        let mut update = LocalUpdate::<K, ConfirmationTimeAnchor>::new(tip.clone());
+
```
I have a feeling that this could be a bit simpler. The algorithm is.

1. Record tip
2. Get relevant transactions anchored at the tip
3. Check tip hasn't changed -- if not extend whichever previous tip still exists in chain with blocks that conflict with those that aren't and finally the new tip.

So I'd put most of this logic at the end. I feel like there might be a more concise way of expressing it.

evanlinjin replied:
Do you mind being a bit more specific about the part you are talking about? Thanks

LLFourn replied:
The logic mentioned in point (3). Right now that happens at the record tip step rather than at the end.

evanlinjin replied:
I've added a separate method for constructing the "update tip" (named `construct_update_tip`). It's probably not what you had in mind @LLFourn, but I think it's the only way of guaranteeing that we create a consistent view of checkpoints during the update.

The logic of finding the point of agreement can screw up consistency, if there is a double-reorg during the update. Please refer to the diagram.

![photo_2023-06-29_13-07-37](https://github.com/bitcoindevkit/bdk/assets/14065705/d3525700-0e18-4fb3-bf30-beaabc9aff3f)

Essentially, we need to atomically fetch the latest x blocks from electrum/esplora to guarantee consistency. x is the assume-final depth.

evanlinjin replied:
Fixed and redesigned in a47837c0a80d9d860b4c5774a978205ca82d7e64

evanlinjin committed:
[bitcoind_rpc] Introduce `prune_and_apply_update` for `IndexedTxGraph`

`prune_and_apply_update` first scans all txs contained in `update`
through the index, then filters out txs using `I::is_tx_relevant` before
applying the update. This is useful for block-by-block syncing.

`Wallet::apply_update` now has a second input; `prune: bool`. If `prune`
is set, irrelevant transactions of `update` will not be included.

evanlinjin committed:
[bitcoind_rpc] Initial work on RPC example

evanlinjin committed:
[local_chain] API simplifications

Revert `get_or_insert` back as `insert_block`.

Method `update` now mutates `LocalChain` directly, instead of mutating
via a second call.

`CheckPoint::new_with_prev` is replaced with `CheckPoint::extend`.

notmandatory commented:
A little off-topic for this PR, but how'd you'all feel about adding type aliases for the block height and unix time instead of just using a `u32` and `u64`? It would make the code easier to read and I don't think there are any drawbacks. Could put them in `chain/src/lib.rs`.

```rust
// The bitcoin block height, 0 is the genesis block.
pub type BlockHeight = u32;

// Number of seconds that have elapsed since 00:00:00 UTC on 1 January 1970, the Unix epoch.
pub type UnixSeconds = u64;
```

evanlinjin committed:
[local_chain] Fix incorrect optimisation logic in `update()`

Within `update()`, it is not always necessary to call `fix_links()`. The
logic to detect this was wrong previously.

Add test that would fail with the previous logic.

evanlinjin committed:
[esplora/electrum] Make updates more consistent

For esplora:

* Separate checkpoint-update logic into a separate method
  `construct_update_tip`.
* We batch-get latest blocks to ensure consistency.
* Reorg-mitigation logic is changed to only reconstruct checkpoints and
  anchors.

For electrum:

* Rename `prepare_chain_update` to `construct_update_tip`. Use
  `Client::block_headers` to get latest headers atomically (instead of
fetching headers one by one).
* `determine_tx_anchor` now uses the lowest anchor checkpoint possible.
* Add comments for better documentation.

evanlinjin commented:
@notmandatory Thank you for the suggestion! Personally, the aliases don't make the code easier to read for me. However, I am open for the change if enough people see the benefit.

notmandatory commented:
I did a little local regtest testing with the `rpc_example` CLI and created a README with my steps: https://github.com/bitcoindevkit/bdk/pull/1019/commits/4d365919d5c797e43040fc20f83f50133a4691c2

@evanlinjin do you think this is useful? If so feel free to pull the README file into your PR or I can add it later in a new PR. For the record I didn't find any problems in this basic manual testing.

evanlinjin committed:
Add more documentation to `LocalChain`

evanlinjin committed:
Add test for `IndexedTxGraph::prune_and_apply_update`

evanlinjin committed:
[bitcoind_rpc] Add docs.

LLFourn commented on crates/chain/src/local_chain.rs:
```diff
@@ -19,215 +130,295 @@ impl ChainOracle for LocalChain {
+        for cp in new_tip.iter() {
+            let block = cp.block_id();
+
+            match self.checkpoints.get(&block.height) {
```
I think this algorithm does not require us to use `BTreeMap`s at all. We are just trying to figure out whether and where two linked lists connect to each other with our given chain rules. This can be done by traversing the linked lists backwards in tandem. The `BTreeMap` should just be there to index the linked list items by height once we've figured out our new chain.

I don't mind implementing it like this if you don't see my vision here.

evanlinjin replied:
@LLFourn today I tried implementing it the way as you suggested. However, the algorithm to traverse the linked lists in tandem is difficult (I will not claim that it is impossible).

One of the reasons that makes this a difficult problem is that our update logic requires a specific rule which states:

> The update chain must have a checkpoint that overlaps the original chain's checkpoint which is one higher than the highest point of agreement.

However, depending on where we have our points of agreements (or lack of), we will need to have to include/exclude specific ranges of blocks from the original and update chains, while ensuring we don't unnecessarily remove links for the final resultant chain.

Essentially, it is much easier to not care about links until the final step (hence the `fix_links` method).

Would you be okay to keep this as is and we fix `LocalChain` for good in #1005?

evanlinjin replied:
@LLFourn has been persistent to have this change (and for good reason) since it has simplified our implementation (see the `local_chain::merge_chains` method). These changes are included in this PR.

LLFourn reviewed (commented):
Comments at this stage:

1. The test look great. They need to also test what happens if you build the update chain from the existing chain local chain. This could be done as new tests and/or having a generic transformation on the existing tests to make one of the update chain nodes be cloned form the same `Arc` as the local chain.
2. I think the `BtreeMap` in `LocalChain` should just index a `CheckPoint` chain. It should not have `CheckPoint`s in it. Will explain more in code comment.
3. There are too many named loops in the bitcoin rpc code. Try to rewrite it so that loop labels are not used as GOTOs.
4. The bitcoin RPC code is using iterators in a kind of unconventional way where you can return `Some(_)` after `None`. I think the idea is cute and works but needs some docs to explain in which situation you would use it as an iterator and which you would manually call `.next` on it.

evanlinjin committed:
`bdk_esplora` redesign

The old logic cannot guarantee consistency of the chain history (i.e. we
may result in a `LocalChain` state which contain blocks that cannot
belong in the same chain). We also had to do a rescan if the tip has
changed since the start of the scan.

To fix this, we anchor transactions at the confirmation height and hash
(which is provided by the esplora API). We fetch the `LocalChain` update
separately, and guarantee consistency by fetching most-recent blocks
atomically via the `GET /blocks` endpoint.

Because the anchors need to be reflected in `LocalChain`,
`TxGraph::missing_blocks(&self, chain: &LocalChain)` is introduced to
fetch missing blocks in a separate call.

evanlinjin committed:
Attempt to make `LocalChain::update` independent of checkpoint map

This logic should carry over better when we change `LocalChain` to be
monotone (#1005).

vladimirfomene commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -0,0 +1,316 @@
+    },
+}
+
+/// A closure that transforms a [`BitcoindRpcItem`] into a [`ConfirmationHeightAnchor`].
```
Update `BitcoindRpcItem` to `BitcoindRpcUpdate`.

vladimirfomene commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -0,0 +1,316 @@
+    }
+}
+
+/// A closure that transforms a [`BitcoindRpcItem`] into a [`ConfirmationTimeAnchor`].
```
Also update here

vladimirfomene commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -0,0 +1,316 @@
+        matches!(self, Self::Block { .. })
+    }
+
+    /// Transforms the [`BitcoindRpcItem`] into a [`LocalUpdate`].
```
Also update `BitcoindRpcItem`

vladimirfomene commented on crates/bitcoind_rpc/src/lib.rs:
```diff
@@ -0,0 +1,316 @@
+}
+
+impl<'a> BitcoindRpcEmitter<'a> {
+    /// Constructs a new [`BitcoindRpcIter`] with the provided [`bitcoincore_rpc::Client`].
```
Update `BitcoindRpcIter` to `BitcoindRpcEmitter`

vladimirfomene commented on crates/chain/src/indexed_tx_graph.rs:
```diff
@@ -90,6 +90,38 @@ where
         }
     }
 
+    /// Apply `update`, but filters out irrelevant transactions.
```
Should pruning irrelevant transactions not be something we do by default (maybe in `apply_update`)? I'm thinking we will save space in doing so.

evanlinjin replied:
For spk-based syncing (such as esplora and electrum) there is nothing to prune. If we enforce pruning, it will interfere with the ability of adding transactions for feerate calculations (for example, a payjoin send).

Via @LLFourn's advice, I've currently split this into two methods: 68b3c5841477d5318e1b144db9d6927a0bf400d4

vladimirfomene commented on crates/chain/src/local_chain.rs:
```diff
@@ -19,215 +144,229 @@ impl ChainOracle for LocalChain {
         &self,
         block: BlockId,
-        static_block: BlockId,
+        chain_tip: BlockId,
```
Sorry if this is was tackled in other PRs. Why don't we just return `Result<bool, Self::Error>`? Is the error here infallible?

evanlinjin replied:
The error here is `Infallible` (as defined 5 lines above). The signature of the trait method `ChainOracle::is_block_in_chain` returns an option wrapped in a result. `Ok(None)` basically means "I don't know". The plan is to have additional logic to handle this specific scenario (not implemented).

vladimirfomene commented on crates/chain/src/local_chain.rs:
```diff
@@ -247,4 +386,185 @@ impl core::fmt::Display for InsertBlockNotMatchingError {
+    let mut o_visited = BTreeMap::<u32, CheckPoint>::new();
+    let mut o_iter = original_tip
+        .iter()
+        .inspect(|cp| {
```
Does this not make more sense as a `for_each`?

evanlinjin replied:
No, as `for_each` consumes the entire iterator.

vladimirfomene commented on crates/chain/src/local_chain.rs:
```diff
@@ -247,4 +386,185 @@ impl core::fmt::Display for InsertBlockNotMatchingError {
+        .peekable();
+
+    let mut u_visited = BTreeMap::<u32, CheckPoint>::new();
+    let u_iter = update_tip.iter().inspect(|cp| {
```
Same here

evanlinjin replied:
https://github.com/bitcoindevkit/bdk/pull/1002/commits/7223adb9f801544fa24180fdabc8f6b159da8c31#r1257138658

vladimirfomene commented on crates/chain/src/local_chain.rs:
```diff
@@ -247,4 +386,185 @@ impl core::fmt::Display for InsertBlockNotMatchingError {
+#[cfg(feature = "std")]
+impl std::error::Error for CannotConnectError {}
+
+fn merge(
```
Please can you add more comments to this function, I'm a bit confused on its inner workings.

evanlinjin replied:
Sorry about that. I've converted this PR back to a draft so people don't get confused. Note that @LLFourn has already rewritten this method with good comments.

vladimirfomene replied:
thanks @evanlinjin !

vladimirfomene reviewed (commented):
My first pass of review. Couldn't finish to the end but will definitely go through it again maybe over the weekend.

LLFourn committed:
Implement simpler chain update algorithm

By iterating backwards over the two chains in tandem to find the
difference between them.

LLFourn committed:
Rethink local chain internals and API

Local chain is a linked list whose heights are indexed in a BTreeMap.

LLFourn committed:
Early exit when there are no update blocks left

LLFourn committed:
impl IntoIterator for CheckPoint

evanlinjin committed:
Add lb_height parameter when merging chains

The lower bound height is the lowest checkpoint the update chain can
introduce to the original chain.

Also add optimisation when we have a perfect connection.

Add `debug_assertions` to `LocalChain` methods that mutate. Add more
`LocalChain` unit tests.

evanlinjin commented on crates/chain/src/local_chain.rs:
```diff
@@ -2,15 +2,137 @@
+            curr = curr.push(block).map_err(|_| self.clone())?;
+        }
+        Ok(curr)
+    }
```
There is already a trait called `Extend` with a method with the exact name and similar name: https://doc.rust-lang.org/std/iter/trait.Extend.html

I think we should rename this to `extend_chain` or something else.

evanlinjin replied:
Renamed to `extend_with_blocks` in 4f327134815cd52789cf13a595927e1eaf1cbfc0

LLFourn replied:
nit I prefer `extend`. iirc `CheckPoint` doesn't implement `Iterator` so this is not really a concern. There are plenty of things in `std` library that have a method `extend` that are not the trait. If you are renaming this then you should rename `push` similarly as well.

evanlinjin committed:
Use `introduce_older_blocks` bool variable for `LocalChain::update`

This replaces using a lower bound height (which is a more complex API).

evanlinjin committed:
Remove `prune` param in `Wallet::apply_update` and use 2 methods instead

The introduced method is `Wallet::prune_and_apply_update`.

evanlinjin committed:
Rename `CheckPoint::extend` to `extend_with_blocks`

So we don't confuse it with `Extend::extend` (trait) that resides in
`core::iter::Extend`.

evanlinjin committed:
Update .gitignore to ignore example db files and remove dup. crate

evanlinjin committed:
Simplify `bdk_bitcoind_rpc` API

Remove the `info` field from the `BitcoindRpcUpdate::Block` variant.

`BitcoindRpcUpdate::into_update` now returns the new `CheckPoint` tip +
the `TxGraph` update (instead of `LocalUpdate`). Consequently, the
`anchor` input is also simplified.

evanlinjin commented:
After a call with @LLFourn, I will make the following changes to this PR before splitting it up:

- [x] `LocalChain::update` should take in a new struct (`local_chain::Update`) that combines the `CheckPoint` input with the `introduce_older_blocks: bool`. This is because the boolean argument is determined by the chain data source, and not by the caller/user.
- [x] Don't prematurely optimize. Remove the same-pointer optimization for `local_chain::merge_chains`.
- [x] Remove `derive(Eq, PartialEq, PartialOrd, Ord)` for `CheckPoint`. We don't know what this means.

Because our PR(s) still need to compile, we will split them as follows:

1. Introduce linked-list `LocalChain` and update esplora/electrum/wallet.
2. Introduce bitcoind rpc crate & example. We will need to add `prune_and_apply_update` for `IndexedTxGraph` and `Wallet` (potentially).
3. Add esplora cli example.

This is the best we can do to ensure that our PRs still compiles at every commit. After the PR-split, I will explore the following:

- [ ] `BlockId` can impl `Anchor` but the semantics should be documented. I.e. it means the tx is confirmed in the anchor block (same block).
- [ ] `BitcoindRpcEmitter` should have separate methods for emitting blocks and mempool.
- [ ] `BitcoindUpdate::into_update`'s input closure should be `anchor_filter_map` which returns an `Option<A>`.

evanlinjin committed:
[local_chain] API improvements and simplifications

Thank you to @LLFourn for the suggestions.

* `LocalChain::update` is now renamed back to `LocalChain::apply_update`
  and takes in a new struct `local_chain::Update` that combines a
`CheckPoint` tip and a `introduce_older_blocks: bool`. This is because
the boolean argument is determined by the chain source and not by the
caller/user.
* We've removed the premature optimization in
  `local_chain::merge_chains`.
* Deriving `PartialEq`, `Eq`, `PartialOrd`, `Ord` for `CheckPoint` does
  not make much sense. They are removed for now.
* Fix: bitcoind_rpc Cargo.toml: wrong bdk_chain version

evanlinjin commented:
Replaced by bitcoindevkit/bdk#1034
